### DISCUSSION TITLE
Hetero Graph where some nodes do not have features vectors

### DISCUSSION THREAD

**Discussion by SHamda** [pyg-team/pytorch_geometric#7936] [Q&A] [answered]
If I have a heterogeneous graph in which some nodes possess feature arrays while others do not, as illustrated in the following graph structure:
```
HeteroData(
  user={ x=[2128, 4] },
  tweet={
    x=[2758, 768],
    y=[2758],
    train_mask=[2758],
    val_mask=[2758],
    test_mask=[2758]
  },
  (user, writes, tweet)={ edge_index=[2, 2758] },
  (word, links, lemma)={ edge_index=[2, 10213] },
  (lemma, links, roots)={ edge_index=[2, 5339] },
  (word, in, tweet)={ edge_index=[2, 32879] },
  (tweet, rev_writes, user)={ edge_index=[2, 2758] },
  (lemma, rev_links, word)={ edge_index=[2, 10213] },
  (roots, rev_links, lemma)={ edge_index=[2, 5339] },
  (tweet, rev_in, word)={ edge_index=[2, 32879] }
)
```
However, when I input this graph into the GNN, it fails to account for the node types "word," "lemma," and "root." I understand that these node types were not explicitly defined. My question is as follows: Can node types be inferred from the edge lists, or is there a technique to incorporate the node types without necessarily providing their feature arrays?

**Comment by rusty1s** [comment, marked as answer]
You would still need to define these node types. Correctly, e.g., via
```python
data["tweet"].num_nodes = ...
```
Afterwards, your GNN expects that there exists node features for these types. You can either add a dummy feature vector, e.g., `torch.ones(num_nodes, 1)` to these node types, or learn a feature representation for them via `torch.nn.Embedding`.
